### ISSUE TITLE
PPI distributed training 收敛问题

### ISSUE THREAD

**Issue by cozilla** [alibaba/euler#198]
单worker收敛情况：
INFO:tensorflow:f1 = 0.5928368, loss = 0.43671727, step = 1920 (2.334 sec)
INFO:tensorflow:f1 = 0.5933435, loss = 0.45692495, step = 1940 (2.293 sec)
INFO:tensorflow:f1 = 0.593776, loss = 0.41395587, step = 1960 (2.331 sec)
INFO:tensorflow:f1 = 0.59379023, loss = 0.43308347, step = 1980 (2.453 sec)
INFO:tensorflow:f1 = 0.59410304, loss = 0.43284538, step = 2000 (2.342 sec)
INFO:tensorflow:**f1 = 0.59446037,** loss = 0.41725546, step = 2020 (2.369 sec)

2 workers收敛就出现了问题，在迭代相同情况下，f1比单worker差很多。
INFO:tensorflow:f1 = 0.52126294, loss = 0.4985747, step = 1855 (2.872 sec)
INFO:tensorflow:f1 = 0.5215076, loss = 0.50285107, step = 1895 (2.710 sec)
INFO:tensorflow:f1 = 0.5229915, loss = 0.4858875, step = 1935 (2.710 sec)
INFO:tensorflow:f1 = 0.5244535, loss = 0.665834, step = 1977 (2.449 sec)
INFO:tensorflow:**f1 = 0.5231028**, loss = 0.57733107, step = 2017 (2.375 sec)

command:
 python -m tf_euler --job_name=worker --task_index=0 --ps_hosts localhost:1999 --worker_hosts=localhost:2000 --model_dir ~/ppi1-512 --data_dir hdfs://localhost:4444/ppi4/ --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 --model graphsage_supervised --mode train --num_epochs 30

**Comment by alinamimi** [comment]
大部分分布式训练在小数据集上，相同step，都会比单机差一些

**Comment by cozilla** [comment]
目前多机版本与单机相比，似乎差别太大了。
在ppi上，2 workers在迭代step=5000后，依然没赶上单机step=500的F1。
并且2 worker training时，step = 2420 ~2800， f1 从0.524逐步下降到0.509。

=== 2 workers training, worker-0 log ===
INFO:tensorflow:f1 = 0.52181345, loss = 0.5427235, step = 2055 (2.820 sec)
INFO:tensorflow:f1 = 0.521434, loss = 0.48875308, step = 2096 (3.008 sec)
INFO:tensorflow:f1 = 0.5216052, loss = 0.5050148, step = 2134 (2.610 sec)
INFO:tensorflow:f1 = 0.52214503, loss = 0.51289356, step = 2175 (2.758 sec)
INFO:tensorflow:f1 = 0.5227455, loss = 0.48939344, step = 2216 (2.911 sec)
INFO:tensorflow:f1 = 0.52287537, loss = 0.47939748, step = 2258 (3.193 sec)
INFO:tensorflow:f1 = 0.52287877, loss = 0.5316295, step = 2299 (2.582 sec)
INFO:tensorflow:f1 = 0.5232997, loss = 0.52680475, step = 2340 (3.137 sec)
INFO:tensorflow:f1 = 0.52445865, loss = 0.48814505, step = 2379 (2.584 sec)
**INFO:tensorflow:f1 = 0.5242681, loss = 0.5601523, step = 2420 (2.878 sec)
INFO:tensorflow:f1 = 0.5208724, loss = 3.1158638, step = 2461 (2.733 sec)
INFO:tensorflow:f1 = 0.5171119, loss = 8.6566725, step = 2500 (3.053 sec)
INFO:tensorflow:f1 = 0.51623684, loss = 4.9730334, step = 2541 (2.624 sec)
INFO:tensorflow:f1 = 0.5139453, loss = 2.3734248, step = 2582 (3.149 sec)
INFO:tensorflow:f1 = 0.5125923, loss = 0.8013828, step = 2623 (2.596 sec)
INFO:tensorflow:f1 = 0.5113203, loss = 0.60114866, step = 2665 (3.008 sec)
INFO:tensorflow:f1 = 0.510471, loss = 0.5361286, step = 2705 (2.940 sec)
INFO:tensorflow:f1 = 0.50996524, loss = 0.5249185, step = 2747 (3.052 sec)
INFO:tensorflow:f1 = 0.51014245, loss = 0.519214, step = 2787 (2.369 sec)
INFO:tensorflow:f1 = 0.5099887, loss = 0.5087555, step = 2825 (2.988 sec)**
INFO:tensorflow:f1 = 0.50988036, loss = 0.5156412, step = 2865 (2.524 sec)
INFO:tensorflow:f1 = 0.5099097, loss = 0.48369977, step = 2904 (3.057 sec)
INFO:tensorflow:f1 = 0.51015365, loss = 0.51790065, step = 2944 (2.411 sec)
INFO:tensorflow:f1 = 0.5102111, loss = 0.51180947, step = 2986 (3.156 sec)
INFO:tensorflow:f1 = 0.5103604, loss = 0.5103795, step = 3027 (2.363 sec)
INFO:tensorflow:f1 = 0.5104654, loss = 0.5005926, step = 3067 (3.171 sec)
INFO:tensorflow:f1 = 0.5108011, loss = 0.51140004, step = 3107 (2.634 sec)
INFO:tensorflow:f1 = 0.5114553, loss = 0.50288934, step = 3146 (2.934 sec)
INFO:tensorflow:f1 = 0.51152354, loss = 0.49354562, step = 3185 (2.318 sec)
INFO:tensorflow:f1 = 0.5120119, loss = 0.48820534, step = 3225 (3.042 sec)
INFO:tensorflow:f1 = 0.51220214, loss = 0.50271523, step = 3266 (2.638 sec)
INFO:tensorflow:f1 = 0.5122398, loss = 0.5088135, step = 3306 (2.940 sec)
INFO:tensorflow:f1 = 0.51264983, loss = 0.4924664, step = 3348 (2.847 sec)
INFO:tensorflow:f1 = 0.5129591, loss = 0.5072672, step = 3389 (2.989 sec)
INFO:tensorflow:f1 = 0.51289433, loss = 0.5042328, step = 3429 (2.716 sec)
INFO:tensorflow:f1 = 0.5131478, loss = 0.49294183, step = 3469 (2.686 sec)
INFO:tensorflow:f1 = 0.5131139, loss = 0.49767852, step = 3510 (2.924 sec)
INFO:tensorflow:f1 = 0.513393, loss = 0.48716915, step = 3551 (2.710 sec)
INFO:tensorflow:f1 = 0.51335394, loss = 0.5007279, step = 3591 (2.864 sec)
INFO:tensorflow:f1 = 0.51363456, loss = 0.49916437, step = 3631 (2.251 sec)
INFO:tensorflow:f1 = 0.5137968, loss = 0.49344862, step = 3672 (2.613 sec)
INFO:tensorflow:f1 = 0.5142387, loss = 0.48688734, step = 3712 (2.588 sec)
INFO:tensorflow:f1 = 0.5149182, loss = 0.46603227, step = 3753 (2.852 sec)
INFO:tensorflow:f1 = 0.51548624, loss = 0.4815021, step = 3793 (2.642 sec)
INFO:tensorflow:f1 = 0.51596403, loss = 0.4830691, step = 3833 (3.062 sec)
INFO:tensorflow:f1 = 0.5167187, loss = 0.5016254, step = 3874 (2.608 sec)
INFO:tensorflow:f1 = 0.51705813, loss = 0.4616519, step = 3914 (2.940 sec)
INFO:tensorflow:f1 = 0.5177795, loss = 0.49353614, step = 3954 (2.395 sec)
INFO:tensorflow:f1 = 0.51822275, loss = 0.49847597, step = 3994 (2.857 sec)
INFO:tensorflow:f1 = 0.51889515, loss = 0.48384005, step = 4035 (2.703 sec)
INFO:tensorflow:f1 = 0.5194043, loss = 0.4700913, step = 4075 (2.701 sec)
INFO:tensorflow:f1 = 0.5198552, loss = 0.48104468, step = 4114 (2.461 sec)
INFO:tensorflow:f1 = 0.51999843, loss = 0.48213488, step = 4153 (2.677 sec)
INFO:tensorflow:f1 = 0.52081096, loss = 0.47503752, step = 4195 (2.737 sec)
INFO:tensorflow:f1 = 0.52146053, loss = 0.47602472, step = 4236 (2.558 sec)
INFO:tensorflow:f1 = 0.52171266, loss = 0.48529187, step = 4275 (2.769 sec)
INFO:tensorflow:f1 = 0.5219716, loss = 0.47574738, step = 4315 (2.708 sec)
INFO:tensorflow:f1 = 0.5222203, loss = 0.49711615, step = 4356 (3.157 sec)
INFO:tensorflow:f1 = 0.5225557, loss = 0.4733794, step = 4395 (2.770 sec)
INFO:tensorflow:f1 = 0.52284604, loss = 0.4752442, step = 4434 (2.765 sec)
INFO:tensorflow:f1 = 0.5234394, loss = 0.46129724, step = 4475 (2.898 sec)
INFO:tensorflow:f1 = 0.52437615, loss = 0.45673576, step = 4514 (2.725 sec)
INFO:tensorflow:f1 = 0.5249175, loss = 0.48285064, step = 4554 (3.103 sec)
INFO:tensorflow:f1 = 0.5256661, loss = 0.4771105, step = 4593 (2.854 sec)
INFO:tensorflow:f1 = 0.52604127, loss = 0.48795867, step = 4635 (3.204 sec)
INFO:tensorflow:f1 = 0.5263397, loss = 0.48526415, step = 4675 (2.896 sec)
INFO:tensorflow:f1 = 0.52657115, loss = 0.47699687, step = 4716 (2.991 sec)
INFO:tensorflow:f1 = 0.5268844, loss = 0.4734469, step = 4757 (2.965 sec)
INFO:tensorflow:f1 = 0.52744967, loss = 0.47701055, step = 4800 (3.123 sec)
INFO:tensorflow:f1 = 0.5278519, loss = 0.47017094, step = 4841 (2.984 sec)
INFO:tensorflow:f1 = 0.5282634, loss = 0.47290823, step = 4881 (3.037 sec)
INFO:tensorflow:f1 = 0.5286765, loss = 0.46298677, step = 4921 (3.047 sec)
INFO:tensorflow:f1 = 0.5288804, loss = 0.47249773, step = 4959 (2.790 sec)


=== 单机 training, worker-0 log ===
INFO:tensorflow:f1 = 0.38325286, loss = 0.6934306, step = 0
INFO:tensorflow:f1 = 0.402751, loss = 0.546466, step = 20 (2.295 sec)
INFO:tensorflow:f1 = 0.41928628, loss = 0.54138327, step = 40 (2.184 sec)
INFO:tensorflow:f1 = 0.4327828, loss = 0.519932, step = 60 (2.232 sec)
INFO:tensorflow:f1 = 0.4471006, loss = 0.501339, step = 80 (2.272 sec)
INFO:tensorflow:f1 = 0.4596003, loss = 0.49800357, step = 100 (2.255 sec)
INFO:tensorflow:f1 = 0.4709569, loss = 0.4846629, step = 120 (2.123 sec)
INFO:tensorflow:f1 = 0.47916007, loss = 0.49257192, step = 140 (2.201 sec)
INFO:tensorflow:f1 = 0.4869985, loss = 0.49359855, step = 160 (2.020 sec)
INFO:tensorflow:f1 = 0.49243146, loss = 0.4901605, step = 180 (2.169 sec)
INFO:tensorflow:f1 = 0.49868804, loss = 0.4718668, step = 200 (2.074 sec)
INFO:tensorflow:f1 = 0.50414944, loss = 0.47391474, step = 220 (2.199 sec)
INFO:tensorflow:f1 = 0.510018, loss = 0.4855024, step = 240 (2.177 sec)
INFO:tensorflow:f1 = 0.51373756, loss = 0.4772194, step = 260 (2.283 sec)
INFO:tensorflow:f1 = 0.5184823, loss = 0.47852275, step = 280 (2.153 sec)
INFO:tensorflow:f1 = 0.5228702, loss = 0.47588718, step = 300 (2.279 sec)
INFO:tensorflow:f1 = 0.52524793, loss = 0.47738498, step = 320 (2.219 sec)
INFO:tensorflow:f1 = 0.5306617, loss = 0.46320868, step = 340 (2.282 sec)
INFO:tensorflow:f1 = 0.5321508, loss = 0.4711943, step = 360 (2.101 sec)
INFO:tensorflow:f1 = 0.5335887, loss = 0.4595451, step = 380 (2.155 sec)
INFO:tensorflow:f1 = 0.5368112, loss = 0.46164337, step = 400 (2.150 sec)
INFO:tensorflow:f1 = 0.5382497, loss = 0.45306793, step = 420 (2.282 sec)
INFO:tensorflow:f1 = 0.5410186, loss = 0.47092143, step = 440 (2.190 sec)
INFO:tensorflow:f1 = 0.54203755, loss = 0.47367412, step = 460 (2.128 sec)
INFO:tensorflow:f1 = 0.5438615, loss = 0.4536673, step = 480 (2.078 sec)
INFO:tensorflow:f1 = 0.54683316, loss = 0.46919686, step = 500 (2.175 sec)
INFO:tensorflow:f1 = 0.54849994, loss = 0.47791272, step = 520 (2.302 sec)
INFO:tensorflow:f1 = 0.5490657, loss = 0.46148834, step = 540 (2.429 sec)
INFO:tensorflow:f1 = 0.5506615, loss = 0.46176457, step = 560 (2.332 sec)

**Comment by alinamimi** [comment]
把单机和分布式的命令发一下？

**Comment by cozilla** [comment]
thanks your quick response! 我把ppi data切成了4份。
单机 :
* python -m tf_euler --job_name=worker --task_index=0 --ps_hosts localhost:1999 --worker_hosts=localhost:2000 --model_dir ~/ppi1-512 --data_dir hdfs://localhost:4444/ppi4/ --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 --model graphsage_supervised --mode train --num_epochs 30

2 workers：
* python -m tf_euler --job_name=worker --task_index=0 --ps_hosts localhost:1999 --worker_hosts=localhost:2000,localhost:2001 --model_dir ~/ppi2-512 --data_dir hdfs://localhost:4444/ppi4/ --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 --model graphsage_supervised --mode train --num_epochs 30

* python -m tf_euler --job_name=worker --task_index=1 --ps_hosts localhost:1999 --worker_hosts=localhost:2000,localhost:2001 --model_dir ~/ppi2-512 --data_dir hdfs://localhost:4444/ppi4/ --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 --model graphsage_supervised --mode train --num_epochs 30

**Comment by zakheav** [comment]
INFO:tensorflow:loss = 0.42785493, step = 4181, f1 = 0.58141744 (4.416 sec)
INFO:tensorflow:loss = 0.42123842, step = 4221, f1 = 0.5815154 (4.539 sec)
INFO:tensorflow:loss = 0.44403493, step = 4261, f1 = 0.5818805 (4.691 sec)
INFO:tensorflow:loss = 0.4359532, step = 4301, f1 = 0.58205503 (4.583 sec)
INFO:tensorflow:loss = 0.4137461, step = 4341, f1 = 0.582418 (4.518 sec)
INFO:tensorflow:loss = 0.43774655, step = 4382, f1 = 0.58273125 (4.564 sec)
INFO:tensorflow:loss = 0.42558494, step = 4421, f1 = 0.58303 (4.503 sec)

保证图数据没有错误分片的条件下，把学习率降低到0.001（默认为0.01），然后跑40个epoch，可以达到0.583的f1，接近单机版本。

**Comment by cozilla** [comment]
谢谢，调整lr/batch的确对收敛有帮助。这个之前也发现了。但是如果单机/多机需要不同参数的话， 会对调试造成一样误导。
另外，euler distributed training 是async training，对吗?

**Comment by alinamimi** [comment]
euler distributed training 是async training
在大数据下面，根本不会去调单机的参数，单机也根本训练不出结果。都是以多机为准，单机最多算是一个指导

**Comment by MeliaLin** [comment]
您好！请问--model_dir里面放的是什么？

**Comment by alinamimi** [comment]
放tensorflow的checkpoint

**Comment by MeliaLin** [comment]
好的，非常感谢！

**Comment by MeliaLin** [comment]
您好！单机伪分布式的命令里为什么没有zookeeper？

**Comment by MeliaLin** [comment]
您好！单机伪分布式的命令里为什么没有zookeeper？

**Comment by MeliaLin** [comment]
您好！单机伪分布式的命令里为什么没有zookeeper？

**Comment by MeliaLin** [comment]
您好！单机伪分布式的命令里为什么没有zookeeper？

**Comment by cozilla** [comment]
> 您好！单机伪分布式的命令里为什么没有zookeeper？

dafault zk_addr is 
```
  tf.flags.DEFINE_string('euler_zk_addr', '127.0.0.1:2181',
                         'Euler ZK registration service.')
  tf.flags.DEFINE_string('euler_zk_path', '/tf_euler',
                         'Euler ZK registration node.')
```

**Comment by MeliaLin** [comment]
感谢您的回复！
127.0.0.1:2181 意思是只需启动单机版的zookeeper就可以吗？

**Comment by cozilla** [comment]
> 感谢您的回复！
> 127.0.0.1:2181 意思是只需启动单机版的zookeeper就可以吗？

我的理解是zk只是为了让worker获得一些meta信息，无所谓是否分布式。

**Comment by MeliaLin** [comment]
好的，非常感谢🙏
